src/simplifiers/LARewriter.cc

```diff
@@ -0,0 +1,111 @@
+        unsigned childrenCount = term.size();
+        if (currentEntry.nextChild < childrenCount) {
+            PTRef nextChild = term[currentEntry.nextChild];
+            ++currentEntry.nextChild;
+            if (done[Idx(logic.getPterm(nextChild).getId())] == 0) {
+                toProcess.push_back(DFSEntry(nextChild));
+            }
+            continue;
+        }
+        // If we are here, we have already processed all children
+        assert(done[Idx(term.getId())] == 0);
+        vec<PTRef> newArgs(childrenCount);
+        bool needsChange = false;
+        for (unsigned i = 0; i < childrenCount; ++i) {
+            auto it = substitutions.find(term[i]);
+            bool childChanged = it != substitutions.end();
+            needsChange |= childChanged;
+            newArgs[i] = childChanged ? it->second : term[i];
+        }
+        PTRef newTerm = needsChange ? logic.insertTerm(term.symb(), newArgs) : currentRef;
+        if (needsChange) {
+            substitutions.insert({currentRef, newTerm});
+        }
+        // The reference "term" has now been possibly invalidated! Do not access it anymore!
+        SymRef symRef = logic.getSymRef(newTerm);
+        if (logic.isIntDiv(symRef) || logic.isIntMod(symRef)) {
+            // check cache first
+            PTRef dividend = logic.getPterm(newTerm)[0];
+            PTRef divisor = logic.getPterm(newTerm)[1];
+            auto it = divModeCache.find({dividend, divisor});
+            bool inCache = (it != divModeCache.end());
+            DivModPair divMod = inCache ? it->second : freshDivModPair(logic);
+            if (not inCache) {
+                divModeCache.insert({{dividend, divisor}, divMod});
+            }
+            PTRef divVar = divMod.div;
+            PTRef modVar = divMod.mod;
+            substitutions[currentRef] = logic.isIntDiv(symRef) ? divVar : modVar;
+            // collect the definitions to add
+            assert(logic.isConstant(divisor));
+            auto divisorVal = logic.getNumConst(divisor);
+            // general case
+            auto upperBound = abs(divisorVal) - 1;
+            // dividend = divVar * dividend + modVar
+            // 0 <= modVar <= |dividend| - 1
+            definitions.push(logic.mkAnd(
+                    logic.mkEq(dividend, logic.mkNumPlus(logic.mkNumTimes(divisor, divVar), modVar)),
+                    logic.mkAnd(
+                            logic.mkNumLeq(logic.getTerm_NumZero(), modVar),
+                            logic.mkNumLeq(modVar, logic.mkConst(upperBound))
+                            )
+                    ));
+        }
+        done[Idx(logic.getPterm(currentRef).getId())] = 1;
+        toProcess.pop_back();
+    }
+    if (substitutions.empty()) {
+        assert(definitions.size() == 0);
+        return root;
+    }
+    auto it = substitutions.find(root);
+    assert(it != substitutions.end());
+    definitions.push(it->second);
+    return logic.mkAnd(definitions);
```

Comment:
While writing this function, I realized it is very close to `Logic::simplifyTree`.
I think it would be good to unify the approach.
One idea would be to implement this as part of `LIA::visit`, but I have a different idea that I like better.
We would have a general rewriter (living outside of `Logic`) which would be rewriting the DAG based on registered plugins. The solver than can customize the rewriter for different theories, but it would also allow users to create their own rewriters.